Uzira.rte/Actors/Brains/Uzira/Uzira.lua

```diff
@@ -0,0 +1,270 @@
+local setupMinionAIMode = function(self, minion)
+	if self.minionsShouldGather then
+		minion.AIMode = Actor.AIMODE_GOTO;
+		minion:ClearMovePath();
+		minion:AddAIMOWaypoint(self);
+	else
+		minion.AIMode = Actor.AIMODE_SENTRY;
+		if math.random() < 0.25 then
```

Comment:
That's wrong though, at least without other changes. If you have
```lua
local rand = math.random();
if rand < 0.25 then
  ...
elseif rand < 0.1 then
  ...
end
```
The 2nd case will never happen, cause the < check is lower for it. Doing a fresh random decouples this part of things.

You're right that it makes the true probability non-obvious, and the most proper thing would be to have each thing as an individual if check (i.e. no elseifs) with individual math.randoms so the probabilities aren't accidentally coupled, but in a case like this the specifics are arbitrary so I think it's fine anyway.

Reply:
I know it's wrong without other changes. That's why I said "and adjust the probabilities accordingly".

Reply:
It's still not really correct though is it?

If you don't do a new random every time you're shooting yourself in the foot if anything lower down has a higher probability than anything higher up (e.g. if the third thing has a < 0.25 check and the 7th thing has a < 0.26 check, the third will never happen if they share the same random).

If you do do a new random every time, it's still not quite right though, is it? Because the randomness isn't affecting independent things, the checks higher up will be more likely to be overwritten, and thus slightly less like to happen.
Let's keep it easy, say you have 10 checks and all of them are math.random < 0.9, only the last will happen 90% of the time, unless you early return, which makes it the same as elseifs again (in which case the lower ones are less likely to happen). So yeah, the probabilities are always non-obvious because they're all overwriting the same thing.

Reply:
That's my point. These are two ways of representing the exact same time.

A probability of
`if math.random() < 0.5 then
    -- do nothing
elseif math.random() < 0.5 then
    doThing();
end`

Is exactly the same as

`if math.random() < 0.25 then
    doThing();
end`

Doing things in this nested way just makes the probabilities extremely unclear (and, furthermore, it's quite inefficient as it requires generated several random numbers, which is a relatively costly operation).

The last elseif in this chain actually has a probability of 0.25%. Which is far, far less than the 10% that a cursory readingf of this code would suggest. In fact, the probability is so absurdly low (this'll only happen once every 400 actors) that I'm convinced it's actually a bug. Which just goes to show that writing this code this way is unwise, because it's caused a bug.

This should be rewritten as follows:

`local random = math.random();
if random < 0.25 then
    minion.AIMode = Actor.AIMODE_PATROL;
elseif random  < 0.35 then
    local target = MovableMan:GetClosestEnemyActor(minion.Team, self.Pos, self.enemySearchRadius, Vector());
    if target then
        minion.AIMode = Actor.AIMODE_GOTO;
        minion:ClearMovePath();
        minion:AddAIMOWaypoint(target);
    end
elseif random  < 0.45 then
    minion.AIMode = Actor.AIMODE_BRAINHUNT;
else
    minion.AIMode = Actor.AIMODE_SENTRY;
end
`

I've adjusted the probabilities here to what I'd assume were the expected probabilities (10%, not 0.25%).

Reply:
I don't think you're right about the last probability, but it's been a while since I've thought about stats, so maybe I'm wrong (probability of choosing the last -> probability of not choosing first * probability of not choosing 2nd * probability of choosing the last -> 0.75 * 0.9 * 0.1 -> 6.75%).
Also, in this particular case I don't think the added expense of a few math.randoms is a big deal, since this isn't run constantly, and there shouldn't be more than a handful of copies of this script running at a time since it's on a brain. But you are right that reusing the same random number is obviously cheaper.

But anyway
1. I see what you're saying and your setup is more obvious. That said, even though it's better, it's still got problems (specifically, it's fragile and its obviousness falls apart if you need your check to go over 1.0), but it's fine for this particular case so I'll change to it
2. We both agree that it would be good to have a clean, widely reusable function for weighted random selection that we always use in these cases, since people can too easily get it wrong if left to their own devices.
This may be best to do it in lua, cause of lua's soft typing (e.g. if someone wanted to choose between a number, a string and a table, it'll be tough to accommodate all that in cpp)

Reply:
Oh, you are correct. I was reversing the probabilities of the stuff further up the chain.

> I see what you're saying and your setup is more obvious. That said, even though it's better, it's still got problems (specifically, it's fragile and its obviousness falls apart if you need your check to go over 1.0), but it's fine for this particular case so I'll change to it

If you write a check that goes over 1 it’s obviously broken, because math.random() never returns a value above 1?

(side note… apparently Lua’s random()l for whatever godforsaken reason, returns a number between 0 and 1 both inclusive. So `if math.random() < 1 then print(“woo”) end` doesn’t always print… 😒)

Reply:
I kept forgetting to mention, at some point maxim put into place for pre5 a replacement of math.random with our internal random, cause math.random had some cases where it was bad at being random (and this lets us use a consistent seed for randomness for multiplayer in the future).

That said,I think our version is also inclusive on both ends haha